File: ios/Sources/Data/Session.swift
import Foundation
import Observation

/// La session du client. Le numéro est l'identité, l'appareil est reconnu
/// quatre-vingt-dix jours, et le jeton vit dans le trousseau.
@MainActor
@Observable
final class Session {
    enum State: Equatable {
        case anonyme
        case attenteCode(phone: String)
        case connecte(phone: String)
    }

    private enum Key {
        static let token = "device.token"
        static let phone = "device.phone"
        static let agency = "device.agency"
        static let language = "app.language"
        static let notifications = "app.notifications"
    }

    let api: API
    private(set) var state: State = .anonyme
    private(set) var deviceToken: String?
    var agencySlug: String {
        didSet { UserDefaults.standard.set(agencySlug, forKey: Key.agency) }
    }
    var language: String {
        didSet {
            UserDefaults.standard.set([language], forKey: "AppleLanguages")
            UserDefaults.standard.set(language, forKey: Key.language)
        }
    }
    var notificationsEnabled: Bool {
        didSet { UserDefaults.standard.set(notificationsEnabled, forKey: Key.notifications) }
    }
    var lastError: String?

    init(api: API = DemoAPI()) {
        self.api = api
        self.agencySlug = UserDefaults.standard.string(forKey: Key.agency) ?? "tca"
        self.language = UserDefaults.standard.string(forKey: Key.language)
            ?? Locale.preferredLanguages.first.map { String($0.prefix(2)) } ?? "fr"
        self.notificationsEnabled = UserDefaults.standard.bool(forKey: Key.notifications)

        // Affordance de test : les captures et les tests d'interface entrent
        // directement, sans passer par le code. Jamais activée en production.
        if ProcessInfo.processInfo.arguments.contains("-connecte") {
            self.deviceToken = "demo-device-token"
            self.state = .connecte(phone: "+216 98 111 222")
            return
        }

        if let token = Keychain.get(Key.token), let phone = Keychain.get(Key.phone) {
            self.deviceToken = token
            self.state = .connecte(phone: phone)
        }
    }

    /// La langue effective pour choisir un libellé traduit venu du serveur.
    var contentLocale: String { language == "zh-Hans" ? "zh" : language }

    func requestCode(phone: String) async {
        lastError = nil
        do {
            try await api.requestCode(agency: agencySlug, phone: phone)
            state = .attenteCode(phone: phone)
        } catch {
            lastError = error.localizedDescription
        }
    }

    func verify(code: String) async {
        guard case .attenteCode(let phone) = state else { return }
        lastError = nil
        do {
            let token = try await api.verifyCode(agency: agencySlug, phone: phone, code: code)
            Keychain.set(token, for: Key.token)
            Keychain.set(phone, for: Key.phone)
            deviceToken = token
            state = .connecte(phone: phone)
        } catch {
            lastError = error.localizedDescription
        }
    }

    func forgetDevice() {
        Keychain.remove(Key.token)
        Keychain.remove(Key.phone)
        deviceToken = nil
        state = .anonyme
    }

    func backToPhone() {
        state = .anonyme
        lastError = nil
    }
}

File: android/app/src/main/java/app/capmedia/visaflow/data/Session.kt
package app.capmedia.visaflow.data

import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import app.capmedia.visaflow.model.MineBundle
import kotlinx.coroutines.launch

/**
 * La session du client. Le numéro est l'identité, l'appareil est reconnu
 * quatre-vingt-dix jours, et le jeton vit dans le magasin chiffré. Le même
 * modèle que ios/Sources/Data/Session.swift, au mot près.
 */
class Session(
    val api: Api,
    private val store: SecureStore,
    private val prefs: android.content.SharedPreferences,
) : ViewModel() {

    sealed interface State {
        data object Anonyme : State
        data class AttenteCode(val phone: String) : State
        data class Connecte(val phone: String) : State
    }

    var state by mutableStateOf<State>(State.Anonyme)
        private set

    var deviceToken by mutableStateOf<String?>(null)
        private set

    var agencySlug by mutableStateOf(prefs.getString(KEY_AGENCY, "tca") ?: "tca")
        private set

    var language by mutableStateOf(prefs.getString(KEY_LANGUAGE, "fr") ?: "fr")
        private set

    var notificationsEnabled by mutableStateOf(prefs.getBoolean(KEY_NOTIFICATIONS, true))
        private set

    var busy by mutableStateOf(false)
        private set

    var lastError by mutableStateOf<String?>(null)

    var mine by mutableStateOf<MineBundle?>(null)
        private set

    init {
        val token = store.read(SecureStore.TOKEN)
        val phone = store.read(SecureStore.PHONE)
        if (!token.isNullOrBlank() && !phone.isNullOrBlank()) {
            deviceToken = token
            state = State.Connecte(phone)
            refresh()
        }
    }

    fun chooseAgency(slug: String) {
        agencySlug = slug.trim().lowercase()
        prefs.edit().putString(KEY_AGENCY, agencySlug).apply()
    }

    fun chooseLanguage(code: String) {
        language = code
        prefs.edit().putString(KEY_LANGUAGE, code).apply()
    }

    fun chooseNotifications(on: Boolean) {
        notificationsEnabled = on
        prefs.edit().putBoolean(KEY_NOTIFICATIONS, on).apply()
    }

    fun requestCode(phone: String) {
        val clean = phone.trim()
        if (clean.length < 6) return
        run {
            lastError = null
            busy = true
            viewModelScope.launch {
                try {
                    api.requestCode(agencySlug, clean)
                    state = State.AttenteCode(clean)
                } catch (e: Exception) {
                    lastError = e.message
                } finally {
                    busy = false
                }
            }
        }
    }

    fun verify(code: String) {
        val phone = (state as? State.AttenteCode)?.phone ?: return
        lastError = null
        busy = true
        viewModelScope.launch {
            try {
                val token = api.verifyCode(agencySlug, phone, code)
                store.write(SecureStore.TOKEN, token)
                store.write(SecureStore.PHONE, phone)
                deviceToken = token
                state = State.Connecte(phone)
                refresh()
            } catch (e: Exception) {
                lastError = e.message
            } finally {
                busy = false
            }
        }
    }

    fun refresh() {
        val token = deviceToken ?: return
        viewModelScope.launch {
            try {
                mine = api.mine(agencySlug, token)
            } catch (e: Exception) {
                lastError = e.message
            }
        }
    }

    /** Se déconnecter efface le jeton, pas seulement l'écran. Un appareil
     *  prêté ne doit pas garder l'accès aux dossiers du précédent client. */
    fun signOut() {
        store.clear()
        deviceToken = null
        mine = null
        state = State.Anonyme
    }

    fun back() {
        if (state is State.AttenteCode) state = State.Anonyme
    }

    companion object {
        private const val KEY_AGENCY = "device.agency"
        private const val KEY_LANGUAGE = "app.language"
        private const val KEY_NOTIFICATIONS = "app.notifications"

        fun factory(context: Context, api: Api): ViewModelProvider.Factory =
            object : ViewModelProvider.Factory {
                @Suppress("UNCHECKED_CAST")
                override fun <T : ViewModel> create(modelClass: Class<T>): T = Session(
                    api = api,
                    store = SecureStore(context),
                    prefs = context.getSharedPreferences("visaflow", Context.MODE_PRIVATE),
                ) as T
            }
    }
}
